File: demo-apps/ios/VoxaTraceDemo/VoxaTraceDemo/Sections/MultiTrack/MultiTrackView.swift
import SwiftUI

/// Multi-track view demonstrating SonixMixer with Config + Factory pattern.
struct MultiTrackView: View {
    @StateObject private var viewModel = MultiTrackViewModel()

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text("Multi-Track")
                .font(.headline)

            Text("Status: \(viewModel.status)")
                .font(.caption)
                .foregroundColor(.secondary)

            // Time display
            Text("\(viewModel.formattedCurrentTime) / \(viewModel.formattedDuration)")
                .font(.body)

            // Seek slider
            seekSlider

            // Backing volume
            backingVolumeSlider

            // Vocal volume
            vocalVolumeSlider

            // Playback controls
            playbackControls

            // Fade buttons
            fadeControls
        }
        .onAppear {
            Task {
                await viewModel.onAppear()
            }
        }
        .onDisappear {
            viewModel.onDisappear()
        }
    }

    private var seekSlider: some View {
        Slider(
            value: Binding(
                get: { viewModel.seekProgress },
                set: { viewModel.seek(to: $0) }
            )
        )
        .disabled(!viewModel.isLoaded)
    }

    private var backingVolumeSlider: some View {
        HStack {
            Text("Backing:")
                .frame(width: 70, alignment: .leading)
            Slider(
                value: Binding(
                    get: { Double(viewModel.backingVolume) },
                    set: { viewModel.setBackingVolume(Float($0)) }
                )
            )
            Text("\(Int(viewModel.backingVolume * 100))%")
                .frame(width: 40)
        }
    }

    private var vocalVolumeSlider: some View {
        HStack {
            Text("Vocal:")
                .frame(width: 70, alignment: .leading)
            Slider(
                value: Binding(
                    get: { Double(viewModel.vocalVolume) },
                    set: { viewModel.setVocalVolume(Float($0)) }
                )
            )
            Text("\(Int(viewModel.vocalVolume * 100))%")
                .frame(width: 40)
        }
    }

    private var playbackControls: some View {
        HStack(spacing: 12) {
            Button("Play") {
                viewModel.play()
            }
            .disabled(!viewModel.isLoaded || viewModel.isPlaying)
            .buttonStyle(.borderedProminent)

            Button("Pause") {
                viewModel.pause()
            }
            .disabled(!viewModel.isPlaying)
            .buttonStyle(.bordered)

            Button("Stop") {
                viewModel.stop()
            }
            .disabled(!viewModel.isLoaded)
            .buttonStyle(.bordered)
        }
    }

    private var fadeControls: some View {
        HStack(spacing: 12) {
            Button("Fade Vocal Down") {
                viewModel.fadeVocalDown()
            }
            .disabled(!viewModel.isLoaded)
            .buttonStyle(.bordered)
            .controlSize(.small)

            Button("Fade Vocal Up") {
                viewModel.fadeVocalUp()
            }
            .disabled(!viewModel.isLoaded)
            .buttonStyle(.bordered)
            .controlSize(.small)
        }
    }
}

/// Backward compatibility alias.
typealias MultiTrackSection = MultiTrackView

#Preview {
    MultiTrackView()
        .padding()
}

File: demo-apps/android/app/src/main/kotlin/com/musicmuni/voxatrace/demo/sections/multitrack/view/MultiTrackView.kt
package com.musicmuni.voxatrace.demo.sections.multitrack.view

import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import com.musicmuni.voxatrace.demo.sections.multitrack.viewmodel.MultiTrackViewModel

/**
 * MultiTrack View - UI for multi-track mixing demo.
 *
 * Demonstrates:
 * - SonixMixer Builder pattern
 * - Multiple audio track mixing
 * - Per-track volume control
 * - Track fade in/out effects
 */
@Composable
fun MultiTrackView(viewModel: MultiTrackViewModel = viewModel()) {
    val context = LocalContext.current

    // Initialize on first composition
    LaunchedEffect(Unit) {
        viewModel.initialize(context)
    }

    // Collect state from ViewModel
    val isPlaying by viewModel.isPlaying.collectAsStateWithLifecycle()
    val currentTimeMs by viewModel.currentTimeMs.collectAsStateWithLifecycle()
    val durationMs by viewModel.durationMs.collectAsStateWithLifecycle()
    val backingVolume by viewModel.backingVolume.collectAsStateWithLifecycle()
    val vocalVolume by viewModel.vocalVolume.collectAsStateWithLifecycle()
    val status by viewModel.status.collectAsStateWithLifecycle()
    val isLoaded by viewModel.isLoaded.collectAsStateWithLifecycle()

    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Text(
            text = "Multi-Track",
            style = MaterialTheme.typography.titleMedium
        )

        Text(text = "Status: $status", style = MaterialTheme.typography.bodySmall)

        // Time display
        Text(
            text = "${MultiTrackViewModel.formatTime(currentTimeMs)} / ${MultiTrackViewModel.formatTime(durationMs)}",
            style = MaterialTheme.typography.bodyMedium
        )

        // Seek slider
        Slider(
            value = if (durationMs > 0) currentTimeMs.toFloat() / durationMs else 0f,
            onValueChange = { fraction -> viewModel.seek(fraction) },
            enabled = isLoaded,
            modifier = Modifier.fillMaxWidth()
        )

        // Backing volume
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Text("Backing:", modifier = Modifier.width(70.dp))
            Slider(
                value = backingVolume,
                onValueChange = { viewModel.setBackingVolume(it) },
                modifier = Modifier.weight(1f)
            )
            Text("${(backingVolume * 100).toInt()}%", modifier = Modifier.width(40.dp))
        }

        // Vocal volume
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Text("Vocal:", modifier = Modifier.width(70.dp))
            Slider(
                value = vocalVolume,
                onValueChange = { viewModel.setVocalVolume(it) },
                modifier = Modifier.weight(1f)
            )
            Text("${(vocalVolume * 100).toInt()}%", modifier = Modifier.width(40.dp))
        }

        // Playback controls
        Row(
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Button(
                onClick = { viewModel.play() },
                enabled = isLoaded && !isPlaying
            ) {
                Text("Play")
            }
            Button(
                onClick = { viewModel.pause() },
                enabled = isPlaying
            ) {
                Text("Pause")
            }
            Button(
                onClick = { viewModel.stop() },
                enabled = isLoaded
            ) {
                Text("Stop")
            }
        }

        // Fade buttons
        Row(
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Button(
                onClick = { viewModel.fadeVocalDown() },
                enabled = isLoaded
            ) {
                Text("Fade Vocal Down")
            }
            Button(
                onClick = { viewModel.fadeVocalUp() },
                enabled = isLoaded
            ) {
                Text("Fade Vocal Up")
            }
        }
    }
}
